File: problems/0011-container-with-most-water/solution.py
"""
Problem Number: 11
Problem Name: Container With Most Water
Difficulty: Medium
Topics: Array, Two Pointers, Greedy
Companies: Amazon, Google, Facebook
"""

from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        """
        Approach: Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        max_water = 0
        left, right = 0, len(height) - 1
        
        while left < right:
            h = min(height[left], height[right])
            w = right - left
            max_water = max(max_water, h * w)
            
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return max_water

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.maxArea([1,8,6,2,5,4,8,3,7])) # 49
    print(solution.maxArea([1,1]))               # 1

File: problems/0011-container-with-most-water/solution.js
/**
 * @problemNumber: 11
 * @problemName: Container With Most Water
 * @difficulty: Medium
 * @topics: Array, Two Pointers, Greedy
 * @companies: Amazon, Google, Facebook
 */

/**
 * Approach: Two Pointers
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *
 * @param {number[]} height
 * @return {number}
 */
function maxArea(height) {
    let maxWater = 0;
    let left = 0;
    let right = height.length - 1;

    while (left < right) {
        const width = right - left;
        const h = Math.min(height[left], height[right]);
        maxWater = Math.max(maxWater, width * h);

        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return maxWater;
}

// Test cases
console.log(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])); // 49
console.log(maxArea([1, 1]));               // 1

module.exports = { maxArea };
